### PR TITLE
Modified on-ramp intent generation

### PR DESCRIPTION
**Pull request by justindg**
This PR addresses #1738 
- All fiat on-ramp provider data will now be maintained in a file named '**onramp_contracts.json**'. Each item will contain the token symbol, contract address, and on-ramp provider (ie Ramp, Wyre, etc)
- FunctionBar button for '**Buy XXX**' now automatically generates label depending on token symbol. This is to prevent bloating the code with multiple if/else statements and corresponding string resources for each kind of token. If a token is not in the known contracts file, it will display '**Buy Crypto**'.

### CODE REVIEW

**Review by JamesSmartCell — CHANGES_REQUESTED** [review]

**Comment by justindg** [comment]
Hi @JamesSmartCell , thanks for the feedback! I've decided to rework the feature a bit, I'll explain the rationale below.

I have opted to create an `OnRampRepository` class which means the Buy feature is not longer tied to the `FunctionButtonBar` class. In the future, this will make it easier to integrate the 'Buy' feature into any Activity or Fragment via injection. Since the feature is supposed to be dynamic, I thought it would be a good idea not to lump it with the "standard functions". This has greatly reduced clutter in the FunctionButtonBar class as well. The `OnRampRepository` class handles everything from reading through the known contracts file to `Intent `generation. This should be easier to extend in the future than the original implementation.

Since I've created it in a way it could be reused easily (and without being dependent on the FunctionButtonBar), the click handler should be through a new interface `BuyCryptoInterface`, which should be implemented in any `Activity `or `Fragment` that needs a Buy feature. It has a sole method of `void handleBuyFunction(Token token);`

I hope this change has relieved some of the pain points mentioned earlier. Please let me know if you have comments or suggestions.

File: app/src/main/java/com/alphawallet/app/widget/FunctionButtonBar.java [hunk 1]
```diff
@@ -592,30 +605,76 @@ private void addTokenScriptFunctions(Map<String, TSAction> availableFunctions, T
      */
     private boolean setupCustomTokenActions()
     {
+        boolean result = false;
         int chainId = token.tokenInfo.chainId;
         String address = token.getAddress();
```
[the remark this answers is not in the document]
Comment (justindg) [reply app/src/main/java/com/alphawallet/app/widget/FunctionButtonBar.java]: Resolved

File: app/src/main/java/com/alphawallet/app/widget/FunctionButtonBar.java [hunk 2]
```diff
@@ -592,30 +605,76 @@ private void addTokenScriptFunctions(Map<String, TSAction> availableFunctions, T
...
+            else
+            {
+                checkKnownContracts(address);
+            }
+            result = true;
+        }
+
+        return result;
+    }
+
```
[the remark this answers is not in the document]
Comment (justindg) [reply app/src/main/java/com/alphawallet/app/widget/FunctionButtonBar.java]: I completely misunderstood this function before, but I have now reverted the change in the new commit.

File: app/src/main/java/com/alphawallet/app/widget/FunctionButtonBar.java [hunk 3]
```diff
@@ -592,30 +605,76 @@ private void addTokenScriptFunctions(Map<String, TSAction> availableFunctions, T
...
+            addFunction(R.string.action_buy, token.tokenInfo.symbol);
+        }
+        else
+        {
+            this.contract = new OnRampContract(token.tokenInfo.symbol);
+            addFunction(R.string.action_buy, context.getString(R.string.crypto));
+        }
+    }
+
+    private OnRampContract findContract(String address)
```
[the remark this answers is not in the document]
Comment (justindg) [reply app/src/main/java/com/alphawallet/app/widget/FunctionButtonBar.java]: Reworked the json file, by using the contract address as the key. Please let me know if this works for you

**Comment by AW-STJ** [comment]
@JamesSmartCell 

Would you be able to add the MixPanel events as well?

We should measure the usage of deeplinks for the different buying options provided

**Comment by AW-STJ** [comment]
Hey Justin, I tested and it works as expected. Waiting for more feedback from team if any.

In the meantime, I have some cosmetic changes.

I want to remove the "from Ramp" from the Buy button for all tokens.
Buy ETH "from Ramp" --> Buy ETH
Buy xDAI "from Ramp" --> Buy xDAI etc.

**Comment by JamesSmartCell** [comment]
@justindg can you have a look at adding the mixpanel events? You need to inject ```AnalyticsServiceType``` into the ERC20detail viewModel, then when the 'buy' button is clicked you just log the event as per ```analyticsService.track(...```

**Comment by JamesSmartCell** [comment]
@jus

> Hi @JamesSmartCell , thanks for the feedback! I've decided to rework the feature a bit, I'll explain the rationale below.
> 
> I have opted to create an `OnRampRepository` class which means the Buy feature is no longer tied to the `FunctionButtonBar` class. In the future, this will make it easier to integrate the 'Buy' feature into any Activity or Fragment via injection. Since the feature is supposed to be dynamic, I thought it would be a good idea not to lump it with the "standard functions". This has greatly reduced clutter in the `FunctionButtonBar ` class, as well as the corresponding` Activities/ViewModels` that have the feature. The `OnRampRepository` class handles everything from reading through the known contracts file to `Intent `generation. This should be easier to extend in the future than the original implementation.
> 
> Since I've created it in a way it could be reused easily (and without being dependent on the FunctionButtonBar), the click handler should be through a new interface `BuyCryptoInterface`, which should be implemented in any `Activity `or `Fragment` that needs a Buy feature. It has a sole method of `void handleBuyFunction(Token token);`
> 
> A few examples that this can be easily integrated would be:
> 
> * Adding a 'Buy Crypto' option via the SettingsFragment
> * Adding a shortcut/icon to buy within the DAppBrowserFragment
> 
> I hope this change has relieved some of the pain points mentioned earlier. Please let me know if you have comments or suggestions.

@justindg Yes this looks very clean now, nice :)

Can you fix the button text as per @AW-STJ's request and add the analytics call to your new OnRampRepo class. Suggest to use something like this:

```
AnalyticsProperties analyticsProperties = new AnalyticsProperties();
analyticsProperties.setData([symbol of token being purchased]);
analyticsService.track(C.AN_USE_ONRAMP, analyticsProperties);
```

Where AN_USE_ONRAMP is just something like 'Use Onramp'. This is just for ease of finding the new data type in the analytics dump.

**Review by JamesSmartCell — APPROVED** [review]

**Pull request merged**